app/views/pages/report.ejs

```diff
@@ -555,4 +737,276 @@
+  }
+
+  const validateTestingAndIsolation = () => {
+    //validate in close contact
+    var hasContact = radioSet(document.getElementsByName("been-in-contact-with"));
+    if (!hasContact) {
+      return [false,"<%= __('Please select whether you have been in close contact with someone who tested positive for COVID-19') %>"];
+    }
+
+    //validate tested
+    var hasTested = radioSet(document.getElementsByName("been-tested"));
+    if (!hasTested) {
+      return [false,"<%= __('Please select whether you have been tested for COVID-19') %>"];
+    }
+
+    //validate isolation
+    var hasIsolation = radioSet(document.getElementsByName("isolation-status"));
+    if (!hasIsolation) {
+      return [false,"<%= __('Please select your isolation status') %>"];
+    }
+
+    return [true, undefined];
+  }
+
+  const validateSymptoms = () => {
+    var hasSymptom = radioSet(document.getElementsByClassName("symptom"));
+    if (hasSymptom) {
+      const validDate = Date.parse(document.getElementById("symptom-start").value);
+      if (!validDate) {
+        return [false,"<%= __('Please add a start date for your symptoms') %>"];
+      }
+    }
+    return [true, undefined];
+  }
+
+  const validateHealth = () => {
+    //health condition
+    const hasHealthConditon = radioSet(document.getElementsByName("diagnosed-other-conditions"));
+    if (!hasHealthConditon) {
+      return [false,"<%= __('Please select if you have a  preexisting health condition') %>"];
+    }
+
+    //smoking
+    const hasSmoking = radioSet(document.getElementsByName("smoking-habits"));
+    if (!hasSmoking) {
+      return [false,"<%= __('Please select your smoking habits') %>"];
+    }
+
+    return [true,undefined];
+  }
+
+  const validationFunctions = [
+    validateDemographicsAndGeography,
+    validateTestingAndIsolation,
+    validateSymptoms,
+    validateHealth
+  ];
+
+  function radioSet(collection) {
+    var re = false;
+    for (var i = 0; i < collection.length; i++) {
+      if (collection[i].checked) {
+        re = true;
+        break;
+      }
+    };
+    return re;
+  }
+</script>
+
+<script>
+  /* Pagination */
+ const numPages = 5;
+ var currentPage = 1;
+
+ /*increase page number by 1 and rerender the form. 
+ An optional validation function can be used to validate the contents of a page 
+ before you can move to the next page
+ */
+ function nextPage(validationFunction) {
+   if (validationFunction) {
+     valid = validationFunction(); //[success, msg]
+      if (valid[0]) {
+        currentPage = currentPage + 1;
+        updatePaginate();
+      } else {
+        //show validation errors to the user here
+        valid[1] ? toastAlert(valid[1]): false;
+      }
+   } else {
+     /* no validation function given */
+    currentPage = currentPage + 1;
+    updatePaginate();
+   }
+ }
+ 
+ //creates an alert for the user displaying given msg 
+ function toastAlert(msg) {
+   var toast = document.createElement("div");
```

Comment:
Couldn't we just add this element to the html with `display: none`? And then update the message and display the element when `toastAlert` is called.

Reply:
yes that's a much better idea than the current implementation. I will do this if we find a way to make the toast alert accessible other wise it can just be replaced with a native browser `alert()`.

Reply:
This has been done